### PR TITLE
Mark DataContainer methods const where possible

### PR DESCRIPTION
**Pull request by rijobro**
Methods should be marked as `const` where possible (so they can be used by `const` objects)

### CODE REVIEW

File: src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp [hunk 1]
```diff
@@ -644,7 +644,7 @@ AcquisitionsFile::take_over(MRAcquisitionData& ac)
 }
 
 unsigned int 
-AcquisitionsFile::items()
+AcquisitionsFile::items() const
```
Comment (evgueni-ovtchinnikov) [inline src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp]: cannot be const:

	unsigned int na = dataset_->getNumberOfAcquisitions();

and `getNumberOfAcquisitions` is not `const`.

File: src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp [hunk 2]
```diff
@@ -654,7 +654,7 @@ AcquisitionsFile::items()
 }
 
 void 
-AcquisitionsFile::get_acquisition(unsigned int num, ISMRMRD::Acquisition& acq)
+AcquisitionsFile::get_acquisition(unsigned int num, ISMRMRD::Acquisition& acq) const
```
Comment (evgueni-ovtchinnikov) [inline src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp]: cannot be `const`: contains

	dataset_->readAcquisition(ind, acq);

and `readAcquisition(ind, acq)` is not `const`.

**Comment by evgueni-ovtchinnikov** [comment]
`number()` calls `items()`, so cannot be `const` either

i would wait for your pr to build on travis - it may reveal more issues with `const`

**Comment by rijobro** [comment]
Ok. The one I really need is `new_data_container`, so I could take a minimalist approach and only do the ones I need (I thought I'd just do them all whilst I was at it).

**Comment by rijobro** [comment]
Closing as this was superceded by #254.

**Comment by rijobro** [comment]
This is fine for Travis (as much as for master). I'm happy for this to be merged if you are. Pretty small one anyway.

File: src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp (hunk 1 above)
Comment (KrisThielemans) [reply src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp]: this works as we're accessing the method via a ptr. The ptr remains `const` (even if the object it points to isn't). A C++ quirck (but in this case a useful one).

**Comment by evgueni-ovtchinnikov** [comment]
i got this error when building sirf:

```
2>C:\Users\wps46139\Documents\GitHub\build\install\include\stir/VectorWithOffset.h(77): error C2664: 'boost::iterators::iterator_adaptor<stir::detail::VectorWithOffset_iter<T>,elemT *,boost::iterators::use_default,boost::iterators::random_access_traversal_tag,boost::iterators::use_default,boost::iterators::use_default>::iterator_adaptor(boost::iterators::iterator_adaptor<stir::detail::VectorWithOffset_iter<elemT>,elemT *,boost::iterators::use_default,boost::iterators::random_access_traversal_tag,boost::iterators::use_default,boost::iterators::use_default> &&)': cannot convert argument 1 from 'const stir::Array<2,float> *const ' to 'stir::Array<2,float> *const &'
2>          with
2>          [
2>              T=stir::Array<2,float>,
2>              elemT=stir::Array<2,float>
2>          ]
2>  C:\Users\wps46139\Documents\GitHub\build\install\include\stir/VectorWithOffset.h(77): note: Conversion loses qualifiers
2>  c:\users\wps46139\documents\github\build\install\include\stir\FullArrayIterator.h(90): note: see reference to function template instantiation 'stir::detail::VectorWithOffset_iter<T>::VectorWithOffset_iter<const T>(const stir::detail::VectorWithOffset_iter<const T> &,boost::iterators::detail::enable_type *)' being compiled
2>          with
2>          [
2>              T=stir::Array<2,float>
2>          ]
2>  c:\users\wps46139\documents\github\build\install\include\stir\FullArrayIterator.h(86): note: see reference to function template instantiation 'stir::detail::VectorWithOffset_iter<T>::VectorWithOffset_iter<const T>(const stir::detail::VectorWithOffset_iter<const T> &,boost::iterators::detail::enable_type *)' being compiled
2>          with
2>          [
2>              T=stir::Array<2,float>
2>          ]
2>  C:\Users\wps46139\Documents\GitHub\SIRF\src\xSTIR\cSTIR\stir_data_containers.cpp(371): note: see reference to function template instantiation 'stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<T>,stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<stir::Array<1,elemT>>,stir::detail::VectorWithOffset_iter<elemT>,elemT,float &,float *>,elemT,float &,float *>::FullArrayIterator<stir::detail::VectorWithOffset_iter<const T>,stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<const stir::Array<1,elemT>>,stir::detail::VectorWithOffset_iter<const float>,elemT,const float &,const float *>,const float&,const float*>(const stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<const T>,stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<const stir::Array<1,elemT>>,stir::detail::VectorWithOffset_iter<const float>,elemT,const float &,const float *>,elemT,const float &,const float *> &,boost::iterators::detail::enable_type *,boost::iterators::detail::enable_type *)' being compiled
2>          with
2>          [
2>              T=stir::Array<2,float>,
2>              elemT=float
2>          ]
2>  C:\Users\wps46139\Documents\GitHub\SIRF\src\xSTIR\cSTIR\stir_data_containers.cpp(371): note: see reference to function template instantiation 'stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<T>,stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<stir::Array<1,elemT>>,stir::detail::VectorWithOffset_iter<elemT>,elemT,float &,float *>,elemT,float &,float *>::FullArrayIterator<stir::detail::VectorWithOffset_iter<const T>,stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<const stir::Array<1,elemT>>,stir::detail::VectorWithOffset_iter<const float>,elemT,const float &,const float *>,const float&,const float*>(const stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<const T>,stir::FullArrayIterator<stir::detail::VectorWithOffset_iter<const stir::Array<1,elemT>>,stir::detail::VectorWithOffset_iter<const float>,elemT,const float &,const float *>,elemT,const float &,const float *> &,boost::iterators::detail::enable_type *,boost::iterators::detail::enable_type *)' being compiled
2>          with
2>          [
2>              T=stir::Array<2,float>,
2>              elemT=float
2>          ]
2>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\memory(439): note: see reference to class template instantiation 'std::_Ptr_base<_Ty>' being compiled
2>          with
2>          [
2>              _Ty=sirf::PETAcquisitionData
2>          ]
2>  C:\Users\wps46139\Documents\GitHub\SIRF\src\xSTIR\cSTIR\include\sirf/cSTIR/stir_data_containers.h(157): note: see reference to class template instantiation 'std::shared_ptr<sirf::PETAcquisitionData>' being compiled
```

**Comment by KrisThielemans** [comment]
ah. quite weird. It probably is a bug in your Visual Studio (seems to be VS 2015), but let's try and figure it out anyway.

First, which STIR version are you on? If it's not the latest, might be worth checking with that first (although none of the related stuff has changed in ages).

If that doesn't resolve it, would you mind removing the new `const` in `PETAcquisitionData` locally, see what happens, and then add them in one-by-one (or whatever strategy you prefer).

**Comment by evgueni-ovtchinnikov** [comment]
i have just pulled and built the latest STIR and have the same error

will try removing const now

**Comment by evgueni-ovtchinnikov** [comment]
removed const from method `dot` and built successfully

**Comment by KrisThielemans** [comment]
entirely weird is it didn't seem to generate the error there.

what about having the method `const`, but as a check comment out the actual implementation? That way we'd see if it's on those lines or something else weird.

**Comment by KrisThielemans** [comment]
My apologies to Microsoft. It isn't an error in VS2015, and they did actually say where the compilation error occured (in `SIRFImageData::dot`). This seems to be due to the following [here](https://github.com/CCPPETMR/SIRF/blob/d093905334b427950a68d36b2f8cbd133d62453e/src/xSTIR/cSTIR/stir_data_containers.cpp#L362)
```C++
#ifdef _MSC_VER
	Image3DF::full_iterator iter;
	Image3DF::const_full_iterator iter_x;
#else
	typename Array<3, float>::const_full_iterator iter;
	typename Array<3, float>::const_full_iterator iter_x;
#endif
```
it needs `const_full_iterator` now (which would always have worked as we're not modifying anything).

@evgueni-ovtchinnikov, not sure why you have the `ifdef`? The latter version happily compiled for me with VS 2015 (and we don't want to go older than that).

There is another problem with this code [here](https://github.com/CCPPETMR/SIRF/blob/d093905334b427950a68d36b2f8cbd133d62453e/src/xSTIR/cSTIR/stir_data_containers.cpp#L361):
```C++
	STIRImageData& x = (STIRImageData&)a_x;
```
It is amazing that this compiles at it throws away the `const` (it goes to show that C-style casts should be avoided). In addition, it would create a seg-fault of `a_x` is another type of `DataContainer`. Instead, we need
```C++
	const STIRImageData& x = dynamic_cast<const STIRImageData&>(a_x);
```
as that'd throw an exception.( I'd hope it's not necessary to have error handling code here, but don't know).

Once that's done, we can get rid of the declarations and just do
```C++
	for (auto iter = data().begin_all(), iter_x = x.data().begin_all();
              ....)
```

There's similar problems in the `axpby` implementation and presumably elsewhere. @evgueni-ovtchinnikov, can you have a look?

**Comment by evgueni-ovtchinnikov** [comment]
thanks for finding the bug! (i wonder how you managed - do not see SIRFImageData in the error message)

`ifdef`: i have VS2013 on my desktop, would not be able to build without it

dynamic cast: just created new issue #281

**Comment by KrisThielemans** [comment]
> I i wonder how you managed

C:\Users\wps46139\Documents\GitHub\SIRF\src\xSTIR\cSTIR\stir_data_containers.cpp(371)
or thereabouts is `SIRFIMageData::dot`... I hadn't noticed either but played around with commenting out stuff.

`ifdef`: if everything is ok with VS2013, then I'd agree to something like it, but if it requires extra work, I'd rather upgrade your VS :-; However, please check then on the [value of _MSC_VER](https://dev.to/yumetodo/list-of-mscver-and-mscfullver-8nd), as VS has moved on. (Not checking the version is something we're guilty of in STIR as well...)

**Pull request merged**